### DISCUSSION TITLE
How to Set Umami Host & Id From Env Variable in Product Server (No SSR)

### DISCUSSION THREAD

**Discussion by elsell** [ijkml/nuxt-umami#120] [Q&A] [answered]
Is it still possible to set the HOST and ID from environment variables in runtimeConfig as shown in #39?  Specifically set via env var when running a built server in production, no SSR.

I'm running nuxt-umami v3.0.2.

I've set up my `nuxt.config.ts` as shown below, but still get the following error in the browser console (I've verified that the env variables are set correctly):

```yaml
[UMAMI]: Your API endpoint is missing or incorrectly configured. Check `host` & `customEndpoint` in module config. 
[3qcYbTLZ.js:49:4735](http://localhost:3000/_nuxt/3qcYbTLZ.js)
Object { hostname: "localhost", language: "en-US", screen: "<redacted>", url: "<redacted>", referrer: "", title: "" }
[3qcYbTLZ.js:50:17](http://localhost:3000/_nuxt/3qcYbTLZ.js)

[UMAMI]: `id` is missing or incorrectly configured. Check module config. 
[3qcYbTLZ.js:49:4735](http://localhost:3000/_nuxt/3qcYbTLZ.js)
Object { hostname: "localhost", language: "en-US", screen: "<redacted>", url: "<redacted>", referrer: "", title: "" }
```

```ts
// nuxt.config.ts

...

runtimeConfig: {
  public: {
      umamiHost: process.env.NUXT_PUBLIC_UMAMI_HOST,
      umamiId: process.env.NUXT_PUBLIC_UMAMI_ID,
  }
}
```

I have other public runtimeConfig that is set from environment variables that works without issue.

_Originally posted by @elsell in https://github.com/ijkml/nuxt-umami/issues/36#issuecomment-2436150249_

**Comment by ijkml** [comment]
Hi @elsell.

Nuxt Umami should automatically pick up `NUXT_UMAMI_HOST` | `NUXT_PUBLIC_UMAMI_HOST` and `NUXT_UMAMI_ID` | `NUXT_PUBLIC_UMAMI_ID` from your `.env`. You don't need to add it to `runtimeConfig`. If the variables exist, then you should check for correctness. The module uses `URL.canParse` to check if `host` is a valid URL.

Hope this helps.

**Reply by elsell** [reply]
Hey @ijkml, thanks a ton for the quick response. 

I'm curious if that behavior is also to be expected when running the production (built) server without SSR? I _do_ notice env vars being pickup up when running `nuxt dev`.

More explicitly, would you expect the output of `nuxt build` to pick up on env vars at runtime? My experience so far has been that env vars will be recognized by `nuxt-umami` at _build_ time, but not at runtime. 

If you do expect this to work during production, I'll be happy to provide more details to hopefully help understand why it doesn't work for me.

_I'm running Nuxt [v3.13.2](https://github.com/nuxt/nuxt/releases/tag/v3.13.2)_

**Reply by ijkml** [reply]
Yes, @elsell. `nuxt-umami` picks up env vars and makes them available at runtime, this is the expected behavior.
I'd be happy to help if something is not working.

**Reply by elsell** [reply]
Thanks for the reply!

It's possible I have something mis-configured, but here is what I'm doing that leads up to `nuxt-umami` not seeing the configuration values:

## Config

### Versions
Operating System: Windows 11 _(Same experience on Ubuntu 22.04)_
`nuxt-umami`: 3.0.2
`nuxt`: 3.13.2
`node`: 20.10.0

### Nuxt Config
```ts
// nuxt.config.ts
export default defineNuxtConfig({
  modules: [
    '@nuxt/ui',
    'nuxt-umami',
    ...
  ],

  ssr: false,

  umami: {
    logErrors: true,
    autoTrack: true,
    ignoreLocalhost: false,
    customEndpoint: process.env.NUXT_PUBLIC_CUSTOM_ENDPOINT ?? '/analytics/api/send',
  },

  ...
}
```

### Runtime Environment Vars
```bash
NUXT_UMAMI_HOST="http://<my_server>"
NUXT_UMAMI_ID="<my_website_id>"
```

## Steps to Reproduce

1. Run `nuxt build` in a session that does not have any environment variables set.
2. Run `node .output/server/index.mjs` in a session that _does_ have the Umami Host and ID env vars set.
3. Open browser console at `http://localhost:3000` and observe the following error messages:
![image](https://github.com/user-attachments/assets/74e8d5c4-c03d-4623-b60b-186fbe5a4405)


## Additional Notes

As mentioned above, I have tried setting the public runtimeConfig using the `NUXT_PUBLIC_UMAMI_XXX` env variables, and I was able to see those values in the browser console when logging the public config. So I know that the environment variables are present in the session, and that Nuxt is able to read them. 

I'm not sure why nuxt-umami wouldn't be able to and I really appreciate your willingness to take a look at this!! :heart:

**Reply by elsell** [reply]
It's also worth noting that I experience the same behavior when running in a Docker container and when running directly from the host operating system.

**Reply by ijkml** [reply, marked as answer]
>     1. Run `nuxt build` in a session that does not have any environment variables set.
> 
>     2. Run `node .output/server/index.mjs` in a session that _does_ have the Umami Host and ID env vars set.

Ah, I see the problem. The module inlines env variables (into `runtimeConfig` if proxy is enabled, else static file). And because Nuxt modules only run **on build**, the variables _have_ to be present when you run `nuxt build`.

**Reply by elsell** [reply]
Okay, that makes sense!

That's too bad, I was hoping I didn't have to do a `docker image build` for each environment, but if that's the only way then I will make it work!

Thank you very much for you time and the answers!

**Reply by ijkml** [reply]
Ah, sorry to hear that.

If you need don't need any of the nev v3 features, you could downgrade to v2.
v2 used runtimeConfig, that is if you've tested runtimeConfig to work with your setup.
